File: Bruteforce/LC/TOP/61.cpp
// Rotate List
// Solution: Circular Linked List and return new head


#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode *rotateRight(ListNode *head, int k)
{
    if (head == nullptr || head->next == nullptr)
    {
        return head; // edge case
    }
    int n = 1;
    ListNode *end = head;
    while (end->next != nullptr)
    {
        end = end->next;
        n++;
    }
    k %= n;
    // 2
    if (k == 0)
        return head;

    // circular
    end->next = head;

    ListNode *t = head; // new tail
    for (int i = 0; i < n - k - 1; ++i)
    {
        t = t->next;
    }
    ListNode *res = t->next;
    t->next = nullptr;
    return res;
}

int main()
{
    return 0;
}

File: Bruteforce/LC/TOP/JS-version/61.js
function ListNode(val, next = null) {
  this.val = val
  this.next = next
}

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
var rotateRight = function (head, k) {
  if (head == null || head.next == null) return head

  let n = 1
  let end = head
  while (end.next != null) {
    end = end.next
    n++
  }

  k %= n
  if (k === 0) return head

  end.next = head

  let t = head
  for (let i = 0; i < n - k - 1; ++i) t = t.next
  let res = t.next
  t.next = null
  return res
}
